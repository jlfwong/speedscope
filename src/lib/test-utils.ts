import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfileGroup} from '../import'
import {exportProfileGroup, importSpeedscopeProfiles} from './file-format'

interface DumpedProfile {
  name: string
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    name: profile.getName(),
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')

  const profileGroup = await importProfileGroup(path.basename(filepath), input)
  if (profileGroup) {
    expect(profileGroup.name).toMatchSnapshot('profileGroup.name')
    expect(profileGroup.indexToView).toMatchSnapshot('indexToView')

    for (let profile of profileGroup.profiles) {
      expect(dumpProfile(profile)).toMatchSnapshot()
    }
  } else {
    fail('Failed to extract profile')
    return
  }

  const profilesWithoutFilename = await importProfileGroup('unknown', input)
  if (profilesWithoutFilename) {
    expect(profilesWithoutFilename.profiles.length).toEqual(profileGroup.profiles.length)
    profilesWithoutFilename.name = profileGroup.name
    for (let i = 0; i < profileGroup.profiles.length; i++) {
      const a = profileGroup.profiles[i]
      const b = profilesWithoutFilename.profiles[i]
      b.setName(a.getName())
    }
    expect(exportProfileGroup(profileGroup)).toEqual(exportProfileGroup(profilesWithoutFilename))
  } else {
    fail('Failed to extract profile when filename was "unknown"')
    return
  }

  const exported = exportProfileGroup(profileGroup)
  const reimportedGroup = importSpeedscopeProfiles(exported)

  expect(reimportedGroup.name).toEqual(profileGroup.name)
  expect(reimportedGroup.profiles.length).toEqual(profileGroup.profiles.length)

  for (let i = 0; i < profileGroup.profiles.length; i++) {
    const profile = profileGroup.profiles[i]
    const reimported = reimportedGroup.profiles[i]

    expect(reimported.getTotalWeight()).toEqual(profile.getTotalWeight())
    expect(dumpProfile(reimported).stacks.join('\n')).toEqual(
      dumpProfile(profile).stacks.join('\n'),
    )
  }

  const reexported = exportProfileGroup(reimportedGroup)
  expect(exported).toEqual(reexported)
}
