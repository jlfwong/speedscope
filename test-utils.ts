import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfile} from './import'

interface DumpedProfile {
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')
  const profile = await importProfile(path.basename(filepath), input)
  if (profile) {
    expect(dumpProfile(profile)).toMatchSnapshot()
  } else {
    fail('Failed to extract profile')
  }
}
