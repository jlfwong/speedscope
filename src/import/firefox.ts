import {Profile, FrameInfo, CallTreeProfileBuilder} from '../lib/profile'
import {getOrInsert} from '../lib/utils'
import {TimeFormatter} from '../lib/value-formatters'

interface Allocations {
  frames: any[]
  sites: any[]
  sizes: any[]
  timestamps: any[]
}

interface Configuration {
  allocationsMaxLogLength: number
  allocationsSampleProbability: number
  bufferSize: number
  sampleFrequency: number
  withAllocations: boolean
  withMarkers: boolean
  withMemory: boolean
  withTicks: boolean
}

interface Lib {
  arch: string
  breakpadId: string
  debugName: string
  debugPath: string
  end: any
  name: string
  offset: number
  path: string
  start: any
}

interface Meta {
  abi: string
  asyncstack: number
  debug: number
  gcpoison: number
  interval: number
  misc: string
  oscpu: string
  platform: string
  processType: number
  product: string
  shutdownTime?: any
  stackwalk: number
  startTime: number
  toolkit: string
  version: number
}

interface PausedRange {
  endTime: number
  reason: string
  startTime: number
}

type Frame = [number] | [number, number | null, number | null, number, number]

interface FrameTable {
  data: Frame[]
  /*
  schema: {
    location: 0
    implementation: 1
    optimizations: 2
    line: 3
    category: 4
  }
  */
}

interface MarkerMeta {
  category: string
  interval: string
  type: string
}
type Marker = [number, number] | [number, number, MarkerMeta]

interface Markers {
  data: Marker[]
  /*
  schema: {
    name: 0
    time: 1
    data: 2
  }
  */
}

type Sample = [number, number, number] | [number, number, number, number, number]

interface Samples {
  data: Sample[]
  /*
  schema: {
    stack: 0
    time: 1
    responsiveness: 2
    rss: 3
    uss: 4
  }
  */
}

export interface StackTable {
  data: [number | null, number][]
  /*
  schema: {
    prefix: 0
    frame: 1
  }
  */
}

export interface Thread {
  frameTable: FrameTable
  markers: Markers
  name: string
  pid: number
  processType: string
  registerTime: number
  samples: Samples
  stackTable: StackTable
  stringTable: string[]
  tid: number
  unregisterTime?: any
}

export interface FirefoxCPUProfile {
  libs: Lib[]
  meta: Meta
  pausedRanges: PausedRange[]
  processes: any[]
  threads: Thread[]
}

export interface FirefoxProfile {
  allocations: Allocations
  configuration: Configuration
  duration: number
  fileType: string
  frames: any[]
  label: string
  markers: any[]
  memory: any[]
  profile: FirefoxCPUProfile
  ticks: any[]
  version: number
}

export function importFromFirefox(firefoxProfile: FirefoxProfile): Profile {
  const cpuProfile = firefoxProfile.profile

  const thread =
    cpuProfile.threads.length === 1
      ? cpuProfile.threads[0]
      : cpuProfile.threads.filter(t => t.name === 'GeckoMain')[0]

  const frameKeyToFrameInfo = new Map<string, FrameInfo>()

  function extractStack(sample: Sample): FrameInfo[] {
    let stackFrameId: number | null = sample[0]
    const ret: number[] = []

    while (stackFrameId != null) {
      const nextStackFrame: [number | null, number] = thread.stackTable.data[stackFrameId]
      const [nextStackId, frameId] = nextStackFrame
      ret.push(frameId)
      stackFrameId = nextStackId
    }
    ret.reverse()
    return ret
      .map(f => {
        const frameData = thread.frameTable.data[f]
        const location = thread.stringTable[frameData[0]]

        const match = /(.*)\s+\((.*?):?(\d+)?\)$/.exec(location)

        if (!match) return null

        if (match[2].startsWith('resource:') || match[2] === 'self-hosted') {
          // Ignore Firefox-internals stuff
          return null
        }

        return getOrInsert(frameKeyToFrameInfo, location, () => ({
          key: location,
          name: match[1]!,
          file: match[2]!,
          line: match[3] ? parseInt(match[3]) : undefined,
        }))
      })
      .filter(f => f != null) as FrameInfo[]
  }

  const profile = new CallTreeProfileBuilder(firefoxProfile.duration)

  let prevStack: FrameInfo[] = []
  for (let sample of thread.samples.data) {
    const stack = extractStack(sample)
    const value = sample[1]

    // Find lowest common ancestor of the current stack and the previous one
    let lcaIndex = -1

    for (let i = 0; i < Math.min(stack.length, prevStack.length); i++) {
      if (prevStack[i] !== stack[i]) {
        break
      }
      lcaIndex = i
    }

    // Close frames that are no longer open
    for (let i = prevStack.length - 1; i > lcaIndex; i--) {
      profile.leaveFrame(prevStack[i], value)
    }

    for (let i = lcaIndex + 1; i < stack.length; i++) {
      profile.enterFrame(stack[i], value)
    }

    prevStack = stack
  }

  profile.setValueFormatter(new TimeFormatter('milliseconds'))
  return profile.build()
}
