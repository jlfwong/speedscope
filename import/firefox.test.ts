import * as fs from 'fs'
import {dumpProfile} from '../test-utils'
import {importFromFirefox} from './firefox'

test('importFromFirefox', () => {
  const input = fs.readFileSync('./sample/profiles/Firefox/59/simple-firefox.json', 'utf8')
  const profile = importFromFirefox(JSON.parse(input))
  expect(dumpProfile(profile)).toMatchSnapshot()
})
