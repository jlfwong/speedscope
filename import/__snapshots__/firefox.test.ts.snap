// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importFromFirefox 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": "file:///Users/jlfwong/code/speedscope/sample/simple.js",
      "key": "a (file:///Users/jlfwong/code/speedscope/sample/simple.js:1)",
      "line": 1,
      "name": "a",
      "selfWeight": 0,
      "totalWeight": 31.992820999992546,
    },
    Frame {
      "col": undefined,
      "file": "file:///Users/jlfwong/code/speedscope/sample/simple.js",
      "key": "b (file:///Users/jlfwong/code/speedscope/sample/simple.js:8)",
      "line": 8,
      "name": "b",
      "selfWeight": 0,
      "totalWeight": 20.971447000018088,
    },
    Frame {
      "col": undefined,
      "file": "file:///Users/jlfwong/code/speedscope/sample/simple.js",
      "key": "d (file:///Users/jlfwong/code/speedscope/sample/simple.js:20)",
      "line": 20,
      "name": "d",
      "selfWeight": 31.992820999992546,
      "totalWeight": 31.992820999992546,
    },
    Frame {
      "col": undefined,
      "file": "file:///Users/jlfwong/code/speedscope/sample/simple.js",
      "key": "c (file:///Users/jlfwong/code/speedscope/sample/simple.js:14)",
      "line": 14,
      "name": "c",
      "selfWeight": 0,
      "totalWeight": 11.021373999974458,
    },
  ],
  "stacks": Array [
    "a;b;d 0.9895330000144895",
    "a;c;d 1.0207879999943543",
    "a;b;d 0.9820370000088587",
    "a;c;d 1.015954000002239",
    "a;b;d 6.999819999997271",
    "a;c;d 2.0036980000149924",
    "a;b;d 0.9814359999727458",
    "a;c;d 2.0017269999952987",
    "a;b;d 1.9998080000223126",
    "a;c;d 3.0192729999835137",
    "a;b;d 4.031898000015644",
    "a;c;d 1.9599339999840595",
    "a;b;d 4.986914999986766",
  ],
}
`;
